backend/api/src/realms.rs

```diff
@@ -25,73 +38,126 @@ impl Realm {
-            })
-            .unwrap_or_default()
+        self.child_keys.iter()
+            .map(|child| &context.realm_tree.realms[&child])
+            .collect()
     }
 }
 
 pub(crate) struct Tree {
     pub(crate) realms: HashMap<u64, Realm>,
-    children: HashMap<u64, Vec<u64>>,
+    from_path: HashMap<String, u64>,
 }
 
 impl Tree {
-    pub(crate) async fn raw_from_db(
-        db: &Pool,
-    ) -> anyhow::Result<impl TryStream<Ok = Realm, Error = impl std::error::Error>> {
-        let row_stream = db.get().await?
-            .query_raw(
-                "select id, name, parent from realms",
-                std::iter::empty(),
-            ).await?;
-
-        Ok(row_stream.map_ok(|row| Realm {
-            key: row.get_key(0),
-            name: row.get(1),
-            parent_key: row.get_key(2),
-        }))
-    }
-
     pub(crate) async fn from_db(db: &Pool) -> anyhow::Result<Self> {
         // We store the nodes of the realm tree in a hash map
         // accessible by the database ID
-        let realms = Self::raw_from_db(db).await?
+        let mut realms = db.get().await?
+            .query_raw(
+                "select id, name, parent, path_segment from realms",
+                std::iter::empty(),
+            ).await?
+            .map_ok(|row| {
+                let key = row.get_key(0);
+                Realm {
+                    key,
+                    name: row.get(1),
+                    parent_key: if key == 0 { None } else { Some(row.get_key(2)) },
+                    path_segment: row.get(3),
+                    child_keys: vec![],
+                }
+            })
             .map_ok(|realm| (realm.key, realm))
             .try_collect::<HashMap<_, _>>().await?;
 
         // With this, and the `parent` member of the `Realm`,
         // we already have quick access to the data of a realm's parent.
-        // To also get to the children quickly we maintain another map.
-        let mut children = <HashMap<_, Vec<_>>>::new();
+        // To also get to the children quickly we maintain a corresponding list
+        // for each realm
+        let keys = realms.values()
+            .filter_map(
+                |realm| realm.parent_key.map(
+                    |parent_key| (realm.key, parent_key)
+                )
+            )
+            .collect::<Vec<_>>();
+        for (key, parent_key) in keys {
+            let parent = realms.get_mut(&parent_key).context("invalid parent")?;
+            parent.child_keys.push(key);
+        }
+
+        // After this point, we should know the tree structure to be valid.
+        // That is, we can now safely panic if we can't find things in our maps/lists;
+        // that's totally a bug in this code, then, not an inconsistency in the db.
+
+        // We also need a map from the full path to the proper realm,
+        // and conversely, we want to cache the full path inside the realm.
+        let from_path = index_by_path(&mut realms);
```

Comment:
Couldn't this whole function and function call be replace by something like:

```rust
let from_path = realms.iter().map(|(key, realm)| (realm.path(), key)).collect()?;
```

Reply:
Yes, but that would walk the tree for every realm, which I wanted to avoid. :thinking:

Might be premature optimization, though. What do you think?

Reply:
I don't think it makes a difference tbh. The asymptotic runtime is certainly not different, right?

Reply:
For every node (n) you touch every node on the path from it to the root (something between 1 and n), vs. you just look at every node once. I would say it's different, the worst case being n^2.

You could argue of course that realm trees probably have a bounded depth; then it wouldn't be different.

Reply:
The `full_path.to_owned()` and `[full_path, &child.path_segment].join("/")` both have a runtime linearly dependent on `full_path`, which -- assuming every path segment is at least one char -- is linearly dependent on the number of descendants. So I don't think it makes a difference.

Reply:
That is a good point. :thinking:

Reply:
Ok, that required some further refactoring because `path` needs the `realms` map, but it should be done now.